File: tests/LocalFunctionTests.cpp
#include <gtest/gtest.h>

#include "clf/LocalFunction.hpp"
#include "clf/LegendrePolynomials.hpp"
#include "clf/Hypercube.hpp"

using namespace clf;

TEST(LocalFunctionTests, Construction) {
  const std::size_t indim = 5;
  const std::size_t outdim = 3;
  const std::size_t maxOrder = 4;
  std::shared_ptr<MultiIndexSet> set = MultiIndexSet::CreateTotalOrder(indim, maxOrder);

  const double delta = 0.1;
  const Eigen::VectorXd xbar = Eigen::VectorXd::Random(indim);
  auto domain = std::make_shared<Hypercube>(xbar-Eigen::VectorXd::Constant(indim, delta), xbar+Eigen::VectorXd::Constant(indim, delta));
  
  auto basis = std::make_shared<LegendrePolynomials>();
  auto vec = std::make_shared<FeatureVector>(set, basis);
  auto mat = std::make_shared<FeatureMatrix>(vec, outdim, domain);

  LocalFunction func(mat);
  EXPECT_EQ(func.InputDimension(), indim);
  EXPECT_EQ(func.OutputDimension(), outdim);
  EXPECT_EQ(func.NumCoefficients(), mat->numBasisFunctions);

  const Eigen::VectorXd x = xbar + delta*Eigen::VectorXd::Random(indim);
  const Eigen::VectorXd coeff = Eigen::VectorXd::Random(func.NumCoefficients());
  const Eigen::VectorXd eval = func.Evaluate(x, coeff);
  EXPECT_EQ(eval.size(), outdim);
  EXPECT_NEAR((eval-mat->ApplyTranspose(x, coeff)).norm(), 0.0, 1.0e-14);

  Eigen::MatrixXi counts = Eigen::MatrixXi::Zero(indim, 2);
  for( std::size_t i=0; i<counts.cols(); ++i ) {
    for( std::size_t j=0; j<3; ++j ) { ++counts(rand()%(indim-1), i); }
  }
  auto linOp = std::make_shared<LinearDifferentialOperator>(counts, outdim);
  const Eigen::MatrixXd diff = func.Derivative(x, coeff, linOp);
  EXPECT_EQ(diff.rows(), outdim);
  EXPECT_EQ(diff.cols(), counts.cols());
  EXPECT_NEAR((diff-mat->ApplyTranspose(x, coeff, linOp)).norm(), 0.0, 1.0e-14);

  for( std::size_t i=0; i<10; ++i ) { EXPECT_TRUE(domain->Inside(func.SampleDomain())); }
}

File: tests/LocalFunctionTests.py
"""! @brief Test the local function (clf::LocalFunction) python interface"""

import unittest

import random

import numpy as np

import PyCoupledLocalFunctions as clf

class TestLocalFunction(unittest.TestCase):
    """! Test the local function (clf::LocalFunction) python interface"""
    
    def test_evaluation(self):
        """! Test the evaluation of a local function"""
        indim = int(5)
        outdim = int(3)
        maxOrder = int(4)
    
        para = clf.Parameters()
        para.Add('InputDimension', indim)
        para.Add('OutputDimension', outdim)
        para.Add('MaximumOrder', maxOrder)

        multiSet = clf.MultiIndexSet(para)
        leg = clf.LegendrePolynomials()
        center = np.array([random.uniform(-1.0, 1.0) for i in range(indim)])
        domain = clf.Hypercube(center-np.array([0.1]*indim), center+np.array([0.1]*indim))

        func = clf.LocalFunction(multiSet, leg, domain, para)
        self.assertEqual(func.InputDimension(), indim)
        self.assertEqual(func.OutputDimension(), outdim)
        self.assertEqual(func.NumCoefficients(), 378)

        x = center + np.array([random.uniform(-1.0, 1.0) for i in range(indim)])
        coeff = np.array([random.uniform(-1.0, 1.0) for i in range(func.NumCoefficients())])
        fx = func.Evaluate(x, coeff)
        self.assertAlmostEqual(np.linalg.norm(fx-func.featureMatrix.ApplyTranspose(x, coeff)), 0.0)

        for i in range(10):
            self.assertTrue(domain.Inside(func.SampleDomain()))
